mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java

```diff
@@ -723,10 +723,14 @@ private void updateHeartbeatTimeout(TaskExecutorID taskExecutorID) {
 
     private void onCacheJobArtifactsOnTaskExecutorRequest(CacheJobArtifactsOnTaskExecutorRequest request) {
         TaskExecutorState state = this.executorStateManager.get(request.getTaskExecutorID());
-        if (state != null) {
+        if (state != null && state.getGateway() != null) {
             TaskExecutorGateway gateway = state.getGateway().join();
-            List<URI> artifacts = jobArtifactsToCache.stream().map(artifactID -> URI.create(artifactID.getResourceID())).collect(Collectors.toList());
-            gateway.cacheJobArtifacts(new CacheJobArtifactsRequest(artifacts));
+            if (gateway != null) {
+                List<URI> artifacts = jobArtifactsToCache.stream().map(artifactID -> URI.create(artifactID.getResourceID())).collect(Collectors.toList());
```

Comment:
nit: shall we save the URI in cache map directly to avoid the remapping for each TE?

Reply:
let me add a todo.